### PR TITLE
sql/gcjob: adopt setting to enable or disable use of delete range

### PR DESCRIPTION
**Pull request by ajwerner**
Backport will address #86580

Release note: None

### CODE REVIEW

**Review by chengxiong-ruan — APPROVED** [review]

File: pkg/sql/importer/import_stmt_test.go [hunk 1]
```diff
@@ -1998,7 +1999,10 @@ func TestFailedImportGC(t *testing.T) {
 		SQLMemoryPoolSize:        256 << 20,
 		ExternalIODir:            baseDir,
 		Knobs: base.TestingKnobs{
-			GCJob: &sql.GCJobTestingKnobs{RunBeforeResume: func(_ jobspb.JobID) error { <-blockGC; return nil }},
+			GCJob: &sql.GCJobTestingKnobs{
+				RunBeforeResume:      func(_ jobspb.JobID) error { <-blockGC; return nil },
+				SkipWaitingForMVCCGC: true,
+			},
```
Comment (Xiang-Gu) [inline pkg/sql/importer/import_stmt_test.go:2005]: why is this changed needed? It seems to me that this change will ensure the gc job uses delRange, whereas previously it does not.

Comment (ajwerner) [reply pkg/sql/importer/import_stmt_test.go:2005]: I was feeling lazy. I made it so if you set this knob then it implies that you should use the new protocol. I needed the new protocol so that the job status matched what the test expected. 

I could have injected the setting instead.

Comment (Xiang-Gu) [reply pkg/sql/importer/import_stmt_test.go:2005]: similarly, I saw that we also changed a few other tests to ensure we use delRange in the gc job. Can you explain to me why? Thanks

Comment (ajwerner) [reply pkg/sql/importer/import_stmt_test.go:2005]: Can you clarify the question? Is the question "why do the tests expect the new protocol"?

File: pkg/sql/gcjob/gcjobnotifier/notifier.go [hunk 2]
```diff
@@ -185,10 +195,16 @@ func (n *Notifier) maybeNotify() {
 	if !zoneConfigUpdated {
 		return
 	}
-	n.notify()
+	n.notifyLocked()
 }
 
 func (n *Notifier) notify() {
```
Comment (Xiang-Gu) [inline pkg/sql/gcjob/gcjobnotifier/notifier.go:201]: Now all call sites of `notify` must not have already acquired the mutex. Tests should be able to verify that, right?

Comment (ajwerner) [reply pkg/sql/gcjob/gcjobnotifier/notifier.go:201]: I added this because of a race detector failure. In practice that call to notify based on the upgrade seemed to not really happen in a racy way. There are no other callers of notify. What's the ask here?

**Review by Xiang-Gu — COMMENTED** [review]
1. I checked that this session variable `storage.mvcc.range_tombstones.enabled` still defaults to `false`. Is this intended?

2. Out of curiosity, what if this session variable is changed from true to false amid a schema change that generates more than 1 GC jobs? Is it possible that some GC jobs uses DelRange and others don't? Is there any potential hard in this? E.g. `ALTER PRIMARY KEY` in old schema changer will schedule two GC jobs -- "GC for temporary index" and "GC for cleanup job for ALTER PRIMAYR KEY".

File: pkg/sql/importer/import_stmt_test.go (hunk 1 above)
Comment (Xiang-Gu) [reply pkg/sql/importer/import_stmt_test.go:2005]: > so that the job status matched what the test expected. 
Then why did the test pass previously, when it didn't use the new protocol?

**Comment by ajwerner** [comment]
I don't think there's any hazard. If somebody turns it on, then they've turned it on and it's possible that the delete range tombstones will be issued. We don't need to provide any promise other than: if a user never turns it on, we never send the new delete range tombstones. Anything about mixed running status values on jobs is a testing consideration and not a correctness concern as far as I'm concerned. 

The storage team decided we didn't have sufficient confidence to enable the new range deletion tombstones by default in 22.2, but wanted a way to experiment. This provides that.

File: pkg/sql/gcjob/gcjobnotifier/notifier.go (hunk 2 above)
Comment (Xiang-Gu) [reply pkg/sql/gcjob/gcjobnotifier/notifier.go:201]: there is no ask here :) it's more of a clarification question. Thanks!

File: pkg/sql/importer/import_stmt_test.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/importer/import_stmt_test.go:2005]: Before this change, the default was to use the new protocol.

Comment (Xiang-Gu) [reply pkg/sql/importer/import_stmt_test.go:2005]: So, is it because your change made the GC job respect the setting `storage.mvcc.range_tombstones.enabled`, and hence those test that previously used the new protocol will no longer use it. Thus, we need to make it use the new protocol by setting the knob here?

**Review by Xiang-Gu — APPROVED** [review]

**Comment by Xiang-Gu** [comment]
Thanks for your patient explanation! LGTM!

File: pkg/sql/importer/import_stmt_test.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/importer/import_stmt_test.go:2005]: Correct.

**Comment by ajwerner** [comment]
TFTR!

bors r+
